reward/staking_manager.go

```diff
@@ -49,40 +51,68 @@ type StakingManager struct {
 	chainHeadSub         event.Subscription
 }
 
+var (
+	once           sync.Once
+	stakingManager *StakingManager
+)
+
+// NewStakingManager creates and returns StakingManager.
+//
+// On the first call, a StakingManager is created with given parameters.
+// From next calls, the existing StakingManager is returned. (Parameters
+// from the next calls will not affect.)
 func NewStakingManager(bc blockChain, gh governanceHelper, db stakingInfoDB) *StakingManager {
-	return &StakingManager{
-		addressBookConnector: newAddressBookConnector(bc, gh),
-		stakingInfoCache:     newStakingInfoCache(),
-		stakingInfoDB:        db,
-		governanceHelper:     gh,
-		blockchain:           bc,
-		chainHeadChan:        make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
+	if bc != nil && gh != nil {
+		// this is only called once
+		once.Do(func() {
+			stakingManager = &StakingManager{
+				addressBookConnector: newAddressBookConnector(bc, gh),
+				stakingInfoCache:     newStakingInfoCache(),
+				stakingInfoDB:        db,
+				governanceHelper:     gh,
+				blockchain:           bc,
+				chainHeadChan:        make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
+			}
+		})
+	} else {
+		logger.Error("unable to set StakingManager", "blockchain", bc, "governanceHelper", gh)
 	}
+
+	return stakingManager
+}
+
+func GetStakingManager() *StakingManager {
+	return stakingManager
 }
 
 // GetStakingInfo returns a corresponding stakingInfo for a blockNum.
-func (sm *StakingManager) GetStakingInfo(blockNum uint64) *StakingInfo {
+func GetStakingInfo(blockNum uint64) *StakingInfo {
+	if stakingManager == nil {
+		logger.Warn("unable to GetStakingInfo; stakingManager is not set")
```

Comment:
Is `Error` better than `Warn?

Is there any case to call this function if stakingMansger is nil in normal situation?

Reply:
When it is called from `Finalize` and staking manager might not be activated, it is normal situation to return `nil`.
Do you think `Error` is better?

Reply:
`NewStakingManager` is called initially. So this function always returns non-nil to `Finalize`.
So I think this case is not normal case.

Reply:
Ok, I agree with you. I applied your comment.